### DISCUSSION TITLE
Create a drag-and-drop report and have the ability to export it

### DISCUSSION THREAD

**Discussion by paula-mamdouh** [directus/directus#22357] [Feature Requests]
### Summary

Let's say I have a database for students, professors, classes, subjects, exams, scores, attendance, etc.

I want to be able to automate the process of generating a report for each student that:

- Lists their subjects for this year
- States their grades
- Demonstrates their total grade, attendance percentage, credit hours, ...
- Show each subject with the professor name, number of conducted exams, ...

Then I want to be able to export that report into a PDF or Excel sheet.

Finally, I should have the ability to provide parameters to the report (ex: date range, student ID, ...) and the report will calculate accordingly.

The requested feature is not for this exact use case (that's just an example). The feature request is more of a data visualization tool embedded into the system.

### Motivation

In order to achieve this use case, I had to explore data visualization solutions (Microsoft Power BI, Google Data Studio, Apache Superset, ...) which require a steep learning curve in order to reach the result of having one report page.

I believe having a simplified version of these systems integrated into Directus will add great value to it.

### Detailed Design

The main object will be "Report". It has:
- Inputs: parameters to configure the report (ex: date range, student ID, ...)
- Building blocks: they're mainly interface elements similar to fields. 
  - They are visual representations. Each building block should have a method to compute its value.
  - Value computation can be defined through a drag-and-drop UI that represents a function (ex: field + field, or SUM(field))
  - Value computation can also be customized using SQL query from the database directly

### Must Have:
- Create a report
- Build a report (by defining its blocks

### Should Have:
- Export option (PDF or Excel)

### Could Have:
- Custom blocks using SQL
- Connecting to external data sources (using REST APIs)

### Won't Have:
- Connecting to external data sources using custom connectors

### Drawbacks

I believe it is quite complex to implement, and I'm not sure if such a use case is very common. 

### Alternatives

N/A

### Adoption Strategy

N/A

**Comment by pdesuze** [comment]
Excellent idea

**Comment by FinnONeill** [comment]
Literally what I'm trying to do right now, hope this gets taken onboard

**Reply by pepeday** [reply]
Hey, did you ever build something like this? How did you go about it?
